lang/cs/Org.Apache.REEF.Network/Elastic/Operators/Logical/ElasticOperator.cs

```diff
@@ -0,0 +1,552 @@
+// specific language governing permissions and limitations
+// under the License.
+
+using System;
+using Org.Apache.REEF.Driver.Task;
+using Org.Apache.REEF.Tang.Interface;
+using Org.Apache.REEF.Network.Elastic.Driver;
+using Org.Apache.REEF.Tang.Exceptions;
+using Org.Apache.REEF.Network.Elastic.Failures;
+using Org.Apache.REEF.Utilities.Logging;
+using System.Globalization;
+using Org.Apache.REEF.Tang.Implementations.Tang;
+using Org.Apache.REEF.Tang.Util;
+using System.Collections.Generic;
+using Org.Apache.REEF.Tang.Implementations.Configuration;
+using Org.Apache.REEF.Network.Elastic.Topology.Logical;
+using Org.Apache.REEF.Network.Elastic.Topology.Logical.Impl;
+using Org.Apache.REEF.Network.Elastic.Config;
+using Org.Apache.REEF.Network.Elastic.Comm;
+using Org.Apache.REEF.Wake.Time.Event;
+using Org.Apache.REEF.Utilities.Attributes;
+using Org.Apache.REEF.Network.Elastic.Failures.Enum;
+using Org.Apache.REEF.Network.Elastic.Topology.Logical.Enum;
+using Org.Apache.REEF.Wake.StreamingCodec.CommonStreamingCodecs;
+
+namespace Org.Apache.REEF.Network.Elastic.Operators.Logical
+{
+    /// <summary>
+    /// Basic implementation for logical operators.
+    /// Each operator is part of a stage and is parametrized by a topology, a failure
+    /// state machine and a checkpoint policy.
+    /// Operators are composed into pipelines.
+    /// Once a pipeline is finalized, tasks can be added to the operator, which
+    /// will in turn add the tasks to the topology and the failure state machine.
+    /// When no more tasks are added, the operator state must be finalized in order to
+    /// schedule the pipeline for execution.
+    /// </summary>
+    [Unstable("0.16", "API may change")]
+    public abstract class ElasticOperator : IFailureResponse, ITaskMessageResponse
+    {
+        private static readonly Logger Log = Logger.GetLogger(typeof(ElasticOperator));
+
+        protected static readonly Dictionary<Type, IConfiguration> CODECMAP = new Dictionary<Type, IConfiguration>()
+        {
+            {
+                typeof(int), StreamingCodecConfiguration<int>.Conf
+                .Set(StreamingCodecConfiguration<int>.Codec, GenericType<IntStreamingCodec>.Class)
+                .Build()
+            },
+            {
+                typeof(int[]), StreamingCodecConfiguration<int[]>.Conf
+                .Set(StreamingCodecConfiguration<int[]>.Codec, GenericType<IntArrayStreamingCodec>.Class)
+                .Build()
+            },
+            {
+                typeof(float), StreamingCodecConfiguration<float>.Conf
+                .Set(StreamingCodecConfiguration<float>.Codec, GenericType<FloatStreamingCodec>.Class)
+                .Build()
+            },
+            {
+                typeof(float[]), StreamingCodecConfiguration<float[]>.Conf
+                .Set(StreamingCodecConfiguration<float[]>.Codec, GenericType<FloatArrayStreamingCodec>.Class)
+                .Build()
```

Comment:
we probably should be able to avoid repetition here -- I just cannot think of an elegant way to do so... maybe,
```C#
private static KeyValuePair<Type, IConfiguration> Codec<TType, TCodec>()
    where TCodec : IStreamingCodec<TType>
{
    return new KeyValuePair<Type, IConfiguration>(
        typeof(TType), StreamingCodecConfiguration<TType>.Conf
            .Set(StreamingCodecConfiguration<TType>.Codec, GenericType<TCodec>.Class)
            .Build());
}

private static Dictionary<TKey, TValue> AsDictionary<TKey, TValue>(
    params KeyValuePair<TKey, TValue>[] values)
{
    return values.ToDictionary(kv => kv.Key, kv => kv.Value);
}

protected static readonly Dictionary<Type, IConfiguration> CODECMAP = AsDictionary(
    Codec<int, IntStreamingCodec>(),
    Codec<int[], IntArrayStreamingCodec>(),
    Codec<float, FloatStreamingCodec>(),
    Codec<float[], FloatArrayStreamingCodec>()
);
```
dunno..

Reply:
Yea I didn't find a better way either. Fixed.